hexgame-rs/src/game/board.rs

```diff
@@ -0,0 +1,300 @@
+/*
+ A board is a set of exagonal cells stacked in a 2D matrix shape.
+ It can be indexed similarly to a matrix with (x, y) coordinates.
+ The (0,0) coordinate starts at the bottom left corner of the board.
+ The x coordinate indexes horizontally (going from left to right)
+ and the y vertically (going from bottom to top)
+ The exagonal cells make so that the board is slanted in a
+ romboidal shape. The slanting goes to the left.
+ ----------
+ '         '
+  '         '
+   '         '
+     ----------
+*/
+
+#[derive(Debug)]
+pub struct Board {
+    pub dim_x: usize,
+    pub dim_y: usize,
+    pub cells: Vec<Vec<Cell>>,
+}
+
+impl Board {
+    pub fn new() -> Self {
+        Self {
+            dim_x: 0,
+            dim_y: 0,
+            cells: vec![vec![]],
+        }
+    }
+
+    pub fn new_from_dims(dim_x: usize, dim_y: usize) -> Self {
+        let mut matrix = Vec::with_capacity(dim_x);
+        for x in 0..dim_x {
+            let mut column = Vec::with_capacity(dim_y);
+            for y in 0..dim_y {
+                column.push(Cell::new(x, y));
+            }
+            matrix.push(column);
+        }
+        Self {
+            dim_x,
+            dim_y,
+            cells: matrix,
+        }
+    }
+
+    pub fn new_from_dim(dim: usize) -> Self {
+        return Board::new_from_dims(dim, dim);
+    }
+
+    fn is_empty(&self) -> bool {
+        self.dim_x == 0 || self.dim_y == 0
+    }
+
+    fn is_valid_square(&self, x: isize, y: isize) -> bool {
+        if self.is_empty() {
+            panic!("The board is empty")
+        }
+        return !(x < 0
+            || y < 0
+            || x >= self.dim_x.try_into().unwrap()
+            || y >= self.dim_y.try_into().unwrap());
+    }
+
+    fn get_neighbours(&self, x: isize, y: isize) -> Vec<Cell> {
+        let mut neighbours: Vec<Cell> = Vec::with_capacity(8);
+        let neighbours_coords: Vec<(isize, isize)> = vec![
+            (x - 1, y),
+            (x + 1, y),
+            (x, y - 1),
+            (x, y + 1),
+            (x - 1, y - 1),
+            (x + 1, y - 1),
+            (x - 1, y + 1),
+            (x + 1, y + 1),
+        ];
+        for (coord_x, coord_y) in neighbours_coords {
+            if self.is_valid_square(coord_x, coord_y) {
+                neighbours.push(
+                    self.cells[usize::try_from(coord_x).unwrap()]
+                        [usize::try_from(coord_y).unwrap()],
+                );
+            }
+        }
+        return neighbours;
+    }
+
+    fn find_next_node_to_visit(visited_nodes: &Vec<Vec<bool>>) -> Option<(usize, usize)> {
+        let dim_x: usize = visited_nodes.len();
+        let dim_y;
+        match visited_nodes.get(0) {
+            Some(row) => dim_y = row.len(),
+            None => panic!("The board is empty: x dimension is zero"),
+        }
+        if dim_y == 0 {
+            panic!("The board is empty: y dimension is zero")
+        }
+        for x in 0..dim_x {
+            for y in 0..dim_y {
```

Comment:
I would like to point out that here we have an iteration over the values of a matrix with  an `early return` condition.
As soon as a non-visited cell is found, the function returns. Skipping the rest of the iterations.

Also, usually, a generator is an object that contains a certain state, if we were to implement it for a matrix of boolean value this will be a class by itself, which might bring some speed-up in performance given that we implement it correctly but also it will add to the complexity of the project. This is unnecessary here in my opinion.